proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,509 @@
+* Proposal: [0004](INF-0004-register-types-and-diagnostics.md)
+* Author(s): [Joshua Batista](https://github.com/bob80905)
+* Sponsor: TBD
+* Status: **Under Consideration**
+* Impacted Project(s): (LLVM)
+* PRs: [#87578](https://github.com/llvm/llvm-project/pull/87578)
+* Issues: [#57886](https://github.com/llvm/llvm-project/issues/57886)
+
+## Introduction
+Resources are often bound to registers of specific register types in HLSL.
+For example:
+```
+RWBuffer<float> rwbuf: register(u0);
+```
+Here we have the resource type (or variable type), `RWBuffer` with a resource 
+element type of `float` being declared as the variable `rwbuf`, and the variable 
+is bound to a register. The register type is `u` and the register number is `0`.
+There's a specific set of valid register types that can be used to bind a given 
+resource in HLSL. When invalid register types are used to bind resources, helpful 
+diagnostics need to be emitted so that the user knows why the resource type is 
+invalid, or why the register type isn't correct. This spec defines the behavior
+of the compiler for whenever an invalid register type is given, or for when a valid
+register type is used to bind a resource type that for whatever reason is 
+incompatible with that register type.
+
+## Motivation
+
+There are several cases in DXC in which diagnostics that are related to
+invalid register types are out of date or invalid. For example, in the case of:
+
+`float b : register(u4);`
+
+an error will be emitted recommending the use of the 'b, c, or i' register
+type. However the 'i' register type is no longer in support, and the 'b' 
+register type is only reserved for resource types that are constant buffers.
+It is worth noting that there is an overloading of the register(...) keyword
+using the 'c' register type to indicate an offset into a constant buffer for
+numeric types only, as opposed to specifying a resource binding.
+Additionally, it is possible the user is unaware that this variable won't 
+actually be used as a resource, but the compiler doesn't communicate that 
+to the user. It would be great for any HLSL developer to immediately have
+answers to the following questions:
+For any valid variable type that needs to be bound as a resource,
+what is the set of valid register types for that variable,
+and what are all the possible diagnostics that can be emitted if the variable
+is given a register type that isn't in the set of valid register types?
+The design below aims to specify the answer to these questions.

```

Comment:
Adding some bullet points to break this up a bit I think that the long sentence may have lost its way:

> It would be great for any HLSL developer to immediately have answers to the following questions:
> * For any valid variable type that needs to be bound as a resource,
>   * what is the set of valid register types for that variable, and 
>   * what are all the possible diagnostics that can be emitted if the variable is given a register type that isn't in the set of valid register types?

I don't think HLSL developers really care about what all the possible diagnostics are. Actually, the more I look at this the less difference I see between the last two bullet points above.

```suggestion
to the user. We should add diagnostics that inform the HLSL developer what register types are valid for that variable.
```

I'm not sure that much more is needed.